File: MonetDB/src/mapi/clients/Mapi.py
import string 
from socket import *
from os import *

trace=		0
interactive=	0

class server:
	def cmd_intern( self, cmd ):
		try:
			self.socket.send(cmd)
			if (trace>0):
				print 'cmd ', cmd
		except IOError:
			print 'IO error '
	
	def result(self):
		result  = ''
		c = self.getchar()
		try:
			while ( c != '\1'):
				result = result + c
				c = self.getchar()
			self.getprompt()
			if (trace>0):
				print result
		except EOFError:
			print 'end of file'
		except error:
			print 'end of file'
			sys.exit(1)
		return result
	
	def getchar(self):
		try:
			c = self.socket.recv(1)
			if (trace>1):
				print c
			return c
		except IOError:
			print 'IO error '
		except error:
			print 'end of file'
			sys.exit(1)
		return ''

	def getprompt(self):
		self.prompt = ''
		c = self.getchar()
		try:
			while ( c != '\1'):
				self.prompt = self.prompt + c
				c = self.getchar()
		except EOFError:
			print 'end of file'
		if (interactive==1):
			print self.prompt
	
	def __init__ ( self, server, port, user ):
		try:
			self.socket = socket(AF_INET, SOCK_STREAM)
			self.socket.connect(server, port)
			self.prompt = ''
		except IOError:
			print 'server refuses access'
	
		self.cmd_intern(user)
		self.result()
		if (trace>0):
			print 'connected ', self.socket

	def disconnect( self ):
		self.result = self.cmd_intern( 'quit;\n' )
		self.socket.close()
		self.socket = 0;

	def cmd( self, cmd ):
		self.cmd_intern(cmd)
		return self.result()

def hostname():
	p = environ['MAPIPORT']
	if (p != ''):
		p = string.splitfields(p,':')[0]
	else:
		p = 'localhost'
	return p

def portnr():
	p = environ['MAPIPORT']
	if (p != ''):
		p = string.splitfields(p,':')[1]
	else:
		p = 50000
	return p


if __name__ == '__main__':
	import fileinput;

	s = server( 'localhost', 50000, 'niels\n')
	fi = fileinput.FileInput()
	sys.stdout.write( s.prompt )
	line= fi.readline()
	while( line != "quit;\n" ):
		res = s.cmd( line )
		print(res);
		sys.stdout.write( s.prompt )
		line = fi.readline() 
	s.disconnect()

File: MonetDB/src/mapi/clients/php/Mapi.php
<?php
# The contents of this file are subject to the MonetDB Public
# License Version 1.0 (the "License"); you may not use this file
# except in compliance with the License. You may obtain a copy of
# the License at
# http://monetdb.cwi.nl/Legal/MonetDBLicense-1.0.html
#
# Software distributed under the License is distributed on an "AS
# IS" basis, WITHOUT WARRANTY OF ANY KIND, either express or
# implied. See the License for the specific language governing
# rights and limitations under the License.
#
# The Original Code is the Monet Database System.
#
# The Initial Developer of the Original Code is CWI.
# Portions created by CWI are Copyright (C) 1997-2004 CWI.
# All Rights Reserved.
?>

<?php

class Mapi {
	function Mapi($server, $port, $user) {
		$this->server = $server;
		$this->port = $port;
		$this->user = $user;
		$this->socket = socket_create(AF_INET, SOCK_STREAM, 0);
		if (!socket_connect($this->socket, $server, $port))
			echo "server refuses access\n";

		$this->buf = "";
		$this->prompt = "";
		$this->cmd_intern("$user\n");
		$this->result();
	}

	function disconnect() {
		$this->cmd_intern("quit;\n");
		socket_close($this->socket);
	}

	function cmd_intern($cmd) {
		socket_write($this->socket, $cmd, strlen($cmd));
	}

	function getstring() {
		$str = "";
		while (!($index = strpos($this->buf, "\001"))) {
			$str .= $this->buf;
			$len = 4096;
			$this->buf = socket_recv($this->socket, $len, 4096, 0);
		}
		$str .= substr($this->buf, 0, $index);
		$this->buf = substr($this->buf, $index + 1, strlen($this->buf));
		return $str;
	}

	function result() {
		$res = $this->getstring();
		$this->getprompt();
		return $res;
	}

	function getprompt() {
		$prompt = $this->getstring();
	}

	function cmd($cmd) {
		$cmd = "$cmd\n";
		$this->cmd_intern($cmd);
		return $this->result();
	}
}
?>
